### PR TITLE
Setup post and basescraper with QuoraProfileScraper

### PR DESCRIPTION
**Pull request by vibhcool**
**This PR is follow-up of PR #1250**
These are the changes so far I have made:-

1) Modify abstract class `BaseScraper` that will be inherited by all scrapers
3) Configure `QuoraProfileScraper` with `BaseScraper` and `Post` : Related Issue #1230
4) Add `Timeline2` class : Related Issue #1231 , #1244
    `Timeline` iterates `MessageEntry` objects. A lot of classes are dependent on `Timeline` , so it can't be configured to iterate `Post` objects. That is why till the time, I `TwitterScraper` works independently and is not configured with `Post` and `BaseScraper`, I have created `Timeline2` class.
5) Configure `Timeline2` with `QuoraProfileScraper` : Related Issue #1231

To test: http://127.0.0.1:9000/api/quoraprofilescraper?query=Vibhor-Verma-5

EDIT 1:  I have added `TODO`'s where I have to make changes.

EDIT 2:
`TODO`s refer to:-

1) the lines which uses Timeline2 as signature. Here signature needs to be fixed after Timeline gets replaced by Timeline2 class. Timeline2 is temporary class.
2) Dummy variables used: the dummy variables stores the parameters that can be fetched. I have used them. Here get-parameters have to be fetched as extra parameters.

EDIT 3: `diff` `Timeline.java` and `Timeline2.java`. Both are very much alike.
### Short description
I have:
- [x] There is a corresponding issue for this pull request.
- [ ] Mentioned the Issue number in the pull request commit message `Fixes #<number> commit message`
- [ ] There is only strictly only one commit per issue.

### For the reviewers
I have:
- [ ] Reviewed this pull request by an authorized contributor.
- [ ] The reviewer is assigned to the pull request.

### CODE REVIEW

**Comment by vibhcool** [comment]
@Orbiter @sudheesh001 @daminisatya @jig08 @mariobehling @singhpratyush @kavithaenair @SKrPl @hemantjadon @Achint08 @SarishiNoHara  @djmgit 
Please review the changes. 
This is follow-up of PR #1250

File: src/org/loklak/harvester/BaseScraper.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
+package org.loklak.harvester;
+
+import java.io.BufferedReader;
+import java.io.InputStreamReader;
+import java.lang.StringBuilder;
+import org.loklak.server.AbstractAPIHandler;
+import org.loklak.data.DAO;
+import java.net.URL;
```
Comment (kavithaenair) [inline src/org/loklak/harvester/BaseScraper.java]: Follow lexical order here as well.

Comment (kavithaenair) [inline src/org/loklak/harvester/BaseScraper.java]: Follow lexical order here as well.

File: src/org/loklak/harvester/Post.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
+package org.loklak.harvester;
+
+import org.json.JSONObject;
+import org.loklak.data.DAO;
+import java.net.URL;
```
Comment (kavithaenair) [inline src/org/loklak/harvester/Post.java]: Here too.

File: src/org/loklak/harvester/Post.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+    public Date getTimestampDate() {
+        return new Date(this.timestamp);
+    }
+
+    //TODO: Set up TwitterTweet before setting this as abstract
+    public void setPostId() { }
+
+    //TODO: Set up TwitterTweet before setting this as abstract
+    public String getPostId() { return new String(); }
```
Comment (kavithaenair) [inline src/org/loklak/harvester/Post.java]: This doesn't look neat. Can you follow as done in line 44.

File: src/org/json/JSONObject.java [hunk 4]
```diff
@@ -488,11 +488,11 @@ public static String doubleToString(double d) {
...
     public Object get(String key) throws JSONException {
         if (key == null) {
-            throw new JSONException("Null key.");
+            //throw new JSONException("Null key.");
         }
         Object object = this.opt(key);
         if (object == null) {
-            throw new JSONException("JSONObject[" + quote(key) + "] not found.");
+            //throw new JSONException("JSONObject[" + quote(key) + "] not found.");
         }
```
Comment (sudheesh001) [inline src/org/json/JSONObject.java]: These are from the `org/json` library. Edits to this are really not recommended.

File: src/org/loklak/api/search/ConsoleService.java [hunk 5]
```diff
@@ -234,12 +235,14 @@ public String getAPIPath() {
...
+        });
+        /*
+        dbAccess.put(Pattern.compile("SELECT +?(.*?) +?FROM +?quoraprofile +?WHERE +?profile ??= ??'(.*?)' ??;"), matcher -> {
+            BaseScraper quoraScrape = new QuoraProfileScraper(matcher.group(2));
             SusiTransfer transfer = new SusiTransfer(matcher.group(1));
-            json.setData(transfer.conclude(json.getData()));
-            return json;
+            return quoraScrape.getData().toJSON(transfer.conclude(json.getData()));
         });
+        */
```
Comment (sudheesh001) [inline src/org/loklak/api/search/ConsoleService.java]: Any reason why this is entirely commented out?

**Comment by sudheesh001** [comment]
Please provide a test link.

**Comment by vibhcool** [comment]
@hemantjadon , I have used `baseUrl` which points to the website which points to website.
`midUrl` points to part(or webpage) of website we want to scrape
`query` is usable at various methods in the program.
I planned to fetch `extra` parameters from request body instead from url string. I have declared it, not used it till now. 
I will add all parameters in request body, as suggested by @singhpratyush , and keep url as a string of `baseUrl` and `midUrl`.
thus, trying to keep everything hastle-free :)

File: src/org/loklak/api/search/ConsoleService.java (hunk 5 above)
Comment (vibhcool) [inline src/org/loklak/api/search/ConsoleService.java]: fixed this

File: src/org/json/JSONObject.java (hunk 4 above)
Comment (vibhcool) [inline src/org/json/JSONObject.java]: fixed this

**Comment by vibhcool** [comment]
@sudheesh001 made all suggested changes
@kavithaenair fixed all syntax errors and codacy errors (some will be fixed in next PRs)

File: src/org/loklak/harvester/Post.java [hunk 6]
```diff
@@ -35,8 +37,16 @@ public void setTimestamp() {
         this.setTimestamp(timestamp);
     }
 
-    //public abstract void getPostId();
+    public Date getTimestampDate() {
+        return new Date(this.timestamp);
+    }
+
+    //TODO: Set up TwitterTweet before setting this as abstract
```
Comment (singhpratyush) [inline src/org/loklak/harvester/Post.java]: Why?

Comment (vibhcool) [inline src/org/loklak/harvester/Post.java]: I have inherited Post here -> https://github.com/loklak/loklak_server/pull/1249/files#diff-c7c5000d9cf0abf8e64f6de0914d35d3
If I declare methods abstract in `Post` right now, everywhere where AbstractObjectEntry is inherited will have to define them. And not all of those child classes are scrapers

File: src/org/loklak/objects/Timeline2.java [hunk 7]
```diff
@@ -0,0 +1,457 @@
...
+/**
+ * A timeline2 is a structure which holds tweet for the purpose of presentation
+ * There is no tweet retrieval method here, just an iterator which returns the tweets in reverse appearing order
+ */
+//TODO: remove MessageEntry
+public class Timeline2 implements Iterable<Post> {
+
+    public static enum Order {
+        CREATED_AT("date"),
+        TIMESTAMP("long"),
```
Comment (Achint08) [inline src/org/loklak/objects/Timeline2.java]: I don't know but why `TIMESTAMP` is assigned as a long data type?

Comment (hemantjadon) [inline src/org/loklak/objects/Timeline2.java]: @Achint08 TimeStamp Is an `Unsigned Int` `U64` thats why. :)

Comment (vibhcool) [inline src/org/loklak/objects/Timeline2.java]: timestamp is no. of sec from 1 jan 1970, used for creating PostId, one can use it to create date from it.

Comment (Achint08) [inline src/org/loklak/objects/Timeline2.java]: Ok ok cool.

File: src/org/loklak/objects/Timeline2.java [hunk 8]
```diff
@@ -0,0 +1,457 @@
+/**
```
Comment (singhpratyush) [inline src/org/loklak/objects/Timeline2.java]: @vibhcool: Could you please point out the things in this class that are different from `Timeline`?

File: src/org/loklak/objects/Timeline2.java [hunk 9]
```diff
@@ -0,0 +1,457 @@
...
+        }
+
+        if (count == 0) return Long.MAX_VALUE;
+        long timeInterval = latest - earliest;
+        long p = 1 + timeInterval / count;
+        return p < 4000 ? p / 4 + 3000 : p;
+    }    
+
+    //TODO: this passes Timeline as argument
+    public void writeToIndex() {
```
Comment (singhpratyush) [inline src/org/loklak/objects/Timeline2.java]: If the plan is to write things to ES index, I think that Timeline should also define the index name.

Also, as the number of scrapers increases, the initialisation of ES node should automatically create indices. How can this be handled?

**Comment by singhpratyush** [comment]
@vibhcool: The test link is not operational. Please check.

File: src/org/loklak/objects/Timeline2.java (hunk 8 above)
Comment (vibhcool) [inline src/org/loklak/objects/Timeline2.java]: this is Timeline class , just that, it iterates `Post` objects instead of `MessageEntry` objects.

File: src/org/loklak/objects/Timeline2.java (hunk 9 above)
Comment (vibhcool) [inline src/org/loklak/objects/Timeline2.java]: I just commented this part. This isn't relevant now but is needed when I get some scrapers start working together. I added TODO to work on this in another PR.

**Comment by vibhcool** [comment]
@singhpratyush updated the test link, please re-review :)

File: src/org/loklak/harvester/Post.java (hunk 6 above)
Comment (singhpratyush) [inline src/org/loklak/harvester/Post.java]: > not all of those child classes are scrapers

But they still can have a unique identifier..

Comment (vibhcool) [inline src/org/loklak/harvester/Post.java]: They doesn't seem to be needed right now. Also some of those classes are unrelated to what I am trying to achieve. So that is why, just to get `TwitterTweet` be subclass of Post, to get results at top, I have made this arrangement. 😅 
See PR #1277

File: src/org/loklak/objects/Timeline2.java (hunk 9 above)
Comment (singhpratyush) [inline src/org/loklak/objects/Timeline2.java]: That is what I am talking about. Once you get more scrapers into work, you'll need to generalise the class for them. One of the requirement for it would be defining the index in which the data would need to get pushed.

**Comment by SKrPl** [comment]
@vibhcool Please look into codacy issues.

**Comment by vibhcool** [comment]
@SKrPl I have fixed the codacy issues. the present 2 issues that are popping up are needed to be kept till refactoring of TwitterScraper. 😅

**Comment by sudheesh001** [comment]
I like the updates in this pull request however I agree with @singhpratyush suggestions regarding the write to index being a separate class so that as more scrapers are added there would be a single general interface that can be used to push to the local elastic search index.

File: src/org/loklak/objects/Timeline2.java (hunk 9 above)
Comment (vibhcool) [inline src/org/loklak/objects/Timeline2.java]: @singhpratyush @sudheesh001 , about the structure of ES index, the existing structure can work with multiple scrapers. Some changes I think that are needed are:- 
1) presently there is `IndexEntry` object that acts as the interface between the ElasticSearch and the Scraper system. There is a need to change it's generic type's superclass to `Post`
2) There can be 2 approaches: - an index per scraper or a Document per scraper . For both the existing structure can work. some lines of code will be needed to be added to set this up. 

2nd point needs some discussion. I will create issue for this point. :)

Comment (vibhcool) [inline src/org/loklak/objects/Timeline2.java]: @singhpratyush @sudheesh001 see this https://github.com/loklak/loklak_server/issues/1290

File: src/org/loklak/harvester/Post.java [hunk 10]
```diff
@@ -35,8 +37,16 @@ public void setTimestamp() {
...
+    public void setPostId() { }
```
Comment (singhpratyush) [inline src/org/loklak/harvester/Post.java]: No parameter in setter method?

File: src/org/loklak/harvester/Post.java [hunk 11]
```diff
@@ -35,8 +37,16 @@ public void setTimestamp() {
...
 
-    //public abstract String setPostId();
+    //TODO: Set up TwitterTweet before setting this as abstract
+    public String getPostId() {
+        return new String();
```
Comment (singhpratyush) [inline src/org/loklak/harvester/Post.java]: Better to `return "";`.

File: src/org/loklak/harvester/Post.java (hunk 10 above)
Comment (vibhcool) [inline src/org/loklak/harvester/Post.java]: https://dzone.com/articles/getter-setter-use-or-not-use

go through the TODO mentioned, it is to be set abstract.

File: src/org/loklak/harvester/Post.java (hunk 11 above)
Comment (vibhcool) [inline src/org/loklak/harvester/Post.java]: doesn't matter , see the `TODO` mentioned

Comment (vibhcool) [inline src/org/loklak/harvester/Post.java]: :)

File: src/org/loklak/harvester/Post.java (hunk 10 above)
Comment (singhpratyush) [inline src/org/loklak/harvester/Post.java]: Even if you set it to `abstract`, it should take some argument to which it can set the ID to.

I still don't understand how setting this `abstract` now will not work but setting it `abstract` after introducing `TwitterScraper` will work. If you don't have a unique identifier for QuoraScraper now, how will you produce it when `TwitterScraper` is introduced?

And I still think that there can be a unique identifier for each Quora profile that is scraped.

File: src/org/loklak/harvester/Post.java (hunk 11 above)
Comment (vibhcool) [inline src/org/loklak/harvester/Post.java]: made the changes

File: src/org/loklak/harvester/Post.java (hunk 10 above)
Comment (vibhcool) [inline src/org/loklak/harvester/Post.java]: made the changes
